File: gtci/11-tree-breadth-first-search/python3/4-minimum-depth-of-a-binary-tree.py
from collections import deque


class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None


class Solution:
    def findDepth(self, root):
        if root is None:
            return 0

        queue = deque()
        queue.append((root, 1))

        while queue:
            node, depth = queue.popleft()

            if node.left is None and node.right is None:
                return depth

            if node.left is not None:
                queue.append((node.left, depth + 1))

            if node.right is not None:
                queue.append((node.right, depth + 1))

        return 0


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    assert sol.findDepth(root) == 2

    root = TreeNode(4)
    root.left = TreeNode(9)
    root.right = TreeNode(7)
    root.left.left = TreeNode(2)
    root.left.right = TreeNode(6)
    assert sol.findDepth(root) == 2

    root = TreeNode(10)
    root.left = TreeNode(5)
    root.right = TreeNode(3)
    root.left.left = TreeNode(7)
    root.right.right = TreeNode(9)
    assert sol.findDepth(root) == 3

    print("All test cases passed.")

File: gtci/11-tree-breadth-first-search/cpp/4-minimum-depth-of-a-binary-tree.cpp
#include <cassert>
#include <iostream>
#include <queue>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  int findDepth(TreeNode *root) {
    if (root == nullptr)
      return 0;
    
    std::queue<TreeNode*> queue;
    queue.push(root);

    int level = 1;
    while (!queue.empty()) {
      const size_t levelSize = queue.size();

      for (size_t i = 0; i < levelSize; i++) {
        TreeNode* node = queue.front();
        queue.pop();

        if (node->left == nullptr && node->right == nullptr)
          return level;
        
        if (node->left != nullptr)
          queue.push(node->left);
        
        if (node->right != nullptr)
          queue.push(node->right);
      }

      level++;
    }

    return 0;
  }
};

int main() {
    Solution sol = Solution();
    
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);
    assert(sol.findDepth(root) == 2);
    
    root = new TreeNode(12);
    root->left = new TreeNode(7);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(9);
    root->right->left = new TreeNode(10);
    root->right->right = new TreeNode(5);
    assert(sol.findDepth(root) == 3);
    
    root = new TreeNode(10);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->right = new TreeNode(15);
    root->right->right->left = new TreeNode(7);
    assert(sol.findDepth(root) == 2);

    std::cout << "All test cases passed." << std::endl;

    delete root;

    return 0;
}
